drivers/net/native_sim_adapt.c

```diff
@@ -0,0 +1,648 @@
+	}
+
+	return ret;
+}
+
+struct nsim_addrinfo_wrap {
+	struct nsim_addrinfo nsim_addrinfo;
+	struct nsim_sockaddr_storage addr_storage;
+	struct addrinfo *addrinfo;
+};
+
+static int addrinfo_to_nsim(struct addrinfo *res,
+			    struct nsim_addrinfo **nsim_res)
+{
+	struct nsim_addrinfo_wrap *nsim_res_wraps;
+	size_t idx_res = 0;
+	size_t n_res = 0;
+	int ret;
+
+	for (struct addrinfo *res_p = res; res_p; res_p = res_p->ai_next) {
+		n_res++;
+	}
+
+	if (n_res == 0) {
+		return 0;
+	}
+
+	nsim_res_wraps = calloc(n_res, sizeof(*nsim_res_wraps));
+	if (!nsim_res_wraps) {
+		return -NSIM_ENOMEM;
+	}
+
+	for (struct addrinfo *res_p = res; res_p; res_p = res_p->ai_next, idx_res++) {
+		struct nsim_addrinfo_wrap *wrap = &nsim_res_wraps[idx_res];
+
+		wrap->addrinfo = res_p;
+
+		wrap->nsim_addrinfo.ai_flags = res_p->ai_flags;
+
+		ret = socket_family_to_nsim(res_p->ai_family, &wrap->nsim_addrinfo.ai_family);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		ret = socket_type_to_nsim(res_p->ai_socktype, &wrap->nsim_addrinfo.ai_socktype);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		ret = socket_proto_to_nsim(res_p->ai_protocol, &wrap->nsim_addrinfo.ai_protocol);
+		if (ret < 0) {
+			goto free_wraps;
+		}
+
+		wrap->nsim_addrinfo.ai_addr =
+			(struct nsim_sockaddr *)&wrap->addr_storage;
+		wrap->nsim_addrinfo.ai_addrlen = sizeof(wrap->addr_storage);
+
+		sockaddr_to_nsim(res_p->ai_addr, res_p->ai_addrlen,
+				 wrap->nsim_addrinfo.ai_addr, &wrap->nsim_addrinfo.ai_addrlen);
+
+		wrap->nsim_addrinfo.ai_canonname =
+			res_p->ai_canonname ? strdup(res_p->ai_canonname) : NULL;
+		wrap->nsim_addrinfo.ai_next = &wrap[1].nsim_addrinfo;
+	}
+
+	nsim_res_wraps[n_res - 1].nsim_addrinfo.ai_next = NULL;
+
+	*nsim_res = &nsim_res_wraps->nsim_addrinfo;
+
+	return 0;
+
+free_wraps:
+	free(nsim_res_wraps);
+
+	return ret;
+}
+
+int nsim_adapt_getaddrinfo(const char *node, const char *service,
+			   const struct nsim_addrinfo *nsim_hints,
+			   struct nsim_addrinfo **nsim_res,
+			   int *system_errno)
+{
+	struct addrinfo hints;
+	struct addrinfo *res = NULL;
+	int ret;
+
+	if (nsim_hints) {
+		hints.ai_flags = nsim_hints->ai_flags;
+
+		ret = socket_family_from_nsim(nsim_hints->ai_family, &hints.ai_family);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSIM_EAI_SYSTEM;
+		}
+
+		ret = socket_type_from_nsim(nsim_hints->ai_socktype, &hints.ai_socktype);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSIM_EAI_SYSTEM;
+		}
+
+		ret = socket_proto_from_nsim(nsim_hints->ai_protocol, &hints.ai_protocol);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSIM_EAI_SYSTEM;
+		}
+	}
+
+	ret = getaddrinfo(node, service,
+			  nsim_hints ? &hints : NULL,
+			  &res);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = addrinfo_to_nsim(res, nsim_res);
+	if (ret < 0) {
+		*system_errno = -ret;
+		return NSIM_EAI_SYSTEM;
+	}
+
+	return ret;
+}
+
+void nsim_adapt_freeaddrinfo(struct nsim_addrinfo *nsim_res)
+{
+	struct nsim_addrinfo_wrap *wrap =
```

Comment:
It feels like the parameter to this function should have been `struct nsim_addrinfo_wrap *wrap` instead of `nsim_res` ? (so you don't need to get its container structure from this pointer)

Reply:
`struct nsim_addrinfo_wrap` (now `nsos_addrinfo_wrap`) is private to this module, so caller has no knowledge about it.